packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts

```diff
@@ -57,15 +63,144 @@ export class FilteringHierarchyDefinition implements RxjsHierarchyDefinition {
     };
   }
 
+  private pathMatcher({
+    identifierFromPath,
+    nodeKey,
+  }: {
+    identifierFromPath: HierarchyNodeIdentifier;
+    nodeKey: GenericNodeKey | InstancesNodeKey;
+  }): boolean | Promise<boolean> {
+    // Check generic node case
+    if (HierarchyNodeIdentifier.isGenericNodeIdentifier(identifierFromPath)) {
+      if (!HierarchyNodeKey.isGeneric(nodeKey)) {
+        return false;
+      }
+      return nodeKey.id === identifierFromPath.id;
+    }
+    if (HierarchyNodeKey.isGeneric(nodeKey)) {
+      return false;
+    }
+
+    if (identifierFromPath.imodelKey && identifierFromPath.imodelKey !== this._imodelAccess.imodelKey) {
+      return false;
+    }
+    // Find matching ids
+    const instanceKeysMatchingIdentifier = nodeKey.instanceKeys.filter((instanceKey) => instanceKey.id === identifierFromPath.id);
+    if (instanceKeysMatchingIdentifier.length === 0) {
+      return false;
+    }
+    // If some of nodes' keys className exactly matches the className of identifier, we can return early
+    if (instanceKeysMatchingIdentifier.some((instanceKey) => instanceKey.className === identifierFromPath.className)) {
+      return true;
+    }
+    return firstValueFrom(
+      from(instanceKeysMatchingIdentifier).pipe(
+        mergeMap((instanceKey) => {
+          const isDerivedFrom = this._imodelAccess.classDerivesFrom(identifierFromPath.className, instanceKey.className);
+          const isDerivedTo = this._imodelAccess.classDerivesFrom(instanceKey.className, identifierFromPath.className);
+          return merge(
+            isDerivedFrom instanceof Promise ? from(isDerivedFrom) : of(isDerivedFrom),
+            isDerivedTo instanceof Promise ? from(isDerivedTo) : of(isDerivedTo),
+          ).pipe(filter((classDerives) => classDerives));
+        }),
+        defaultIfEmpty(false),
+        take(1),
+      ),
+    );
+  }
+
+  private setAutoExpandForProcessedNode<T extends ProcessedGenericHierarchyNode | ProcessedInstanceHierarchyNode>(
+    node: T,
+    parentNode?: ParentHierarchyNode,
+  ): Observable<T> {
+    const filteringHelper = createHierarchyFilteringHelper(this._nodeIdentifierPaths, parentNode);
+    const nodeAutoExpandPropPossiblyPromise = filteringHelper.createChildNodeProps({
+      asyncPathMatcher: (identifier): boolean | Promise<boolean> => this.pathMatcher({ identifierFromPath: identifier, nodeKey: node.key }),
+      parentKeys: node.parentKeys,
+    });
+    return (nodeAutoExpandPropPossiblyPromise instanceof Promise ? from(nodeAutoExpandPropPossiblyPromise) : of(nodeAutoExpandPropPossiblyPromise)).pipe(
+      map((autoExpandProp) => {
+        if (autoExpandProp?.autoExpand) {
+          Object.assign(node, { autoExpand: true });
+        }
+        return node;
+      }),
+    );
+  }
+
+  private getFilteringPathsForGroupingNode(
```

Comment:
I think this function could easily cause perf issues - we could have 1k+ `parentPaths` and 10k `childNodeIdentifiers`, resulting in 10M iterations.